File: android/src/main/java/expo/modules/wallet/ExpoWalletModule.kt
package expo.modules.wallet

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.net.URL

class ExpoWalletModule : Module() {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  override fun definition() = ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ExpoWallet')` in JavaScript.
    Name("ExpoWallet")

    // Defines a JavaScript synchronous function that runs the native code on the JavaScript thread.
    Function("hello") {
      "Hello world! 👋"
    }
    
      // Function to add passes to Google Wallet using a JWT or save URL
      // Accepts either a single string (JWT/URL) or array of strings for cross-platform compatibility
      AsyncFunction("addPasses") { data: Any ->
        val context = appContext.reactContext
        if (context == null) {
          throw Exception("No React context available")
        }

        val googleWalletData = when (data) {
          is String -> data
          is List<*> -> {
            // If array is passed (iOS-style), take first element as Google Wallet data
            if (data.isEmpty()) {
              throw Exception("Empty pass data provided")
            }
            data[0] as? String ?: throw Exception("Invalid pass data format")
          }
          else -> throw Exception("Invalid data type. Expected String or Array<String>")
        }
      
        // Google Wallet Save Passes intent
        // See: https://developers.google.com/wallet/retail/implement-save-passes
        val intent = android.content.Intent(android.content.Intent.ACTION_VIEW)
        intent.data = android.net.Uri.parse(googleWalletData)
        intent.setPackage("com.google.android.gms")
      
        // Start the intent
        context.startActivity(intent)
        mapOf("added" to 1)
      }

  }
}

File: ios/ExpoWalletModule.swift
import ExpoModulesCore
import PassKit


public class ExpoWalletModule: Module {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  public func definition() -> ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ExpoWallet')` in JavaScript.
    Name("ExpoWallet")

    // Defines a JavaScript synchronous function that runs the native code on the JavaScript thread.
    Function("hello") {
      return "Hello world! 👋"
    }
      

    // Expose addPasses method to JS, accepts pkpass file URI
    // Accepts an array of pkpass URIs
    AsyncFunction("addPasses") { (pkpassUris: [String], promise: Promise) in
      var passes: [PKPass] = []
      for uri in pkpassUris {
        guard let url = URL(string: uri), url.isFileURL else {
          promise.reject("ERR_INVALID_URI", "Invalid pkpass URI: \(uri)")
          return
        }
        do {
          let passData = try Data(contentsOf: url)
          let pass = try PKPass(data: passData)
          passes.append(pass)
        } catch {
          promise.reject("ERR_PASS_LOAD", "Failed to load pass: \(uri), \(error.localizedDescription)")
          return
        }
      }
      let passLibrary = PKPassLibrary()
      passLibrary.addPasses(passes) { status in
        switch status {
        case .didAddPasses:
          promise.resolve(["added": passes.count])
        case .shouldReviewPasses:
          promise.resolve(["review": true, "added": passes.count])
        case .didCancelAddPasses:
          promise.reject("ERR_ADD_PASS_CANCELLED", "User cancelled adding passes.")
        @unknown default:
          promise.reject("ERR_ADD_PASS_UNKNOWN", "Unknown status.")
        }
      }
    }

  }
}
